MsCorePkg/Library/MuSecureBootKeySelectorLib/MuSecureBootKeySelectorLib.c

```diff
@@ -17,9 +17,228 @@
+  Microsoft-serviced keys.
+
+  This function checks the certificates in the provided Secure Boot database (DbVar)
+  against a predefined list of Microsoft certificates.
+  If all certificates in the database match the Microsoft-serviced certificates,
+  the machine is considered `Microsoft Serviced.` If any certificate does not match,
+  the machine is considered `CUSTOM,` and the function returns an error.
+
+  @param[in]  DbVar               Pointer to the Secure Boot database (db) variable.
+  @param[in]  DbVarSize           Size of the Secure Boot database in bytes.
+  @param[out] IsMsServicedConfig  Pointer to a BOOLEAN that will be set to TRUE if
+                                  the database matches Microsoft-serviced keys, or
+                                  FALSE otherwise.
+
+  @retval EFI_SUCCESS             The operation completed successfully, and the result
+                                  is stored in IsMsServicedConfig.
+  @retval EFI_NOT_FOUND           A certificate in the database does not match any
+                                  Microsoft-serviced certificates.
+  @retval EFI_UNSUPPORTED         The hashing protocol or service binding protocol
+                                  required for the operation is not available.
+  @retval EFI_OUT_OF_RESOURCES    Memory allocation failed during the operation.
+  @retval Other                   An error occurred during the operation.
+
+**/
+EFI_STATUS
+EFIAPI
+IsMsftServicedConfig (
+  UINT8    *DbVar,
+  UINTN    DbVarSize,
+  BOOLEAN  *IsMsServicedConfig
+  )
+{
+  EFI_STATUS                    Status;
+  EFI_SIGNATURE_LIST            *CertList;
+  EFI_SIGNATURE_DATA            *Cert;
+  UINTN                         CertCount;
+  UINT8                         *RootCert;
+  UINTN                         RootCertSize;
+  UINTN                         Index;
+  UINTN                         CertIndex;
+  EFI_SERVICE_BINDING_PROTOCOL  *Hash2ServiceBinding;
+  EFI_HASH2_PROTOCOL            *Hash2Protocol;
+  EFI_HASH2_OUTPUT              HashOutput;
+
+  if (IsMsServicedConfig != NULL) {
+    *IsMsServicedConfig = FALSE;
+  }
+
+  Status = gBS->LocateProtocol (&gEfiHash2ProtocolGuid, NULL, (VOID **)&Hash2Protocol);
+  if (EFI_ERROR (Status)) {
+    //
+    // If we can't find the Hashing protocol, then we need to create one.
+    //
+
+    //
+    // Platform is expected to publish the hash service binding protocol.
+    //
+    Status = gBS->LocateProtocol (
+                    &gEfiHash2ServiceBindingProtocolGuid,
+                    NULL,
+                    (VOID **)&Hash2ServiceBinding
+                    );
+    if (EFI_ERROR (Status) || (Hash2ServiceBinding == NULL) || (Hash2ServiceBinding->CreateChild == NULL)) {
+      Status = EFI_UNSUPPORTED;
+      goto Exit;
+    }
+
+    //
+    // Create an instance of the hash protocol for this controller.
+    //
+    Status = Hash2ServiceBinding->CreateChild (Hash2ServiceBinding, &mHash2ServiceHandle);

```

Comment:
If a child instance is created on this handle, you still need to open the protocol on that handle to set `Hash2Protocol` in this case, right?